### ISSUE TITLE
Some test specs trigger compiler bug which eats warnings

### ISSUE THREAD

**Issue by blast-hardcheese** [scalatest/scalatest#2056]
In attempting to track down why an obviously invalid construction was a runtime error instead of a type error, I stumbled upon https://github.com/scala/bug/issues/12441 .

A minimal reproduction is included in [the linked repo](https://github.com/blast-hardcheese/scala-bug-xlint-infer-any), but I'll additionally talk about a workaround here.

The core issue seems to come from when values that are attempting to emit a warning are coerced to `Unit`, wherein their warnings are suppressed:

```scala
object Bogus {
  Option.empty[String].contains(1234) // Emits a warning

  def should(func: => Unit): Unit = func

  should {
    Option.empty[String].contains(1234) // Does not warn
  }
}
```

An interesting observation is that this only happens at the block site, so a perfectly valid workaround for this is to apply:

```diff
diff --git a/src/main/scala/Bogus.scala b/src/main/scala/Bogus.scala
index a94fb91..7304326 100644
--- a/src/main/scala/Bogus.scala
+++ b/src/main/scala/Bogus.scala
@@ -1,10 +1,10 @@
 object Bogus {
   Option.empty[String].contains(1234) // Emits a warning
 
-  def should(func: => Unit): Unit = func
+  def should[A](func: => A): Unit = func
 
   should {
-    Option.empty[String].contains(1234) // Does not warn
+    Option.empty[String].contains(1234) // This now emits a warning!
   }
 
   println(Option.empty[String].contains(1234)) // does not warn
```

Impacted specs
---

I copied the different spec examples from [`Selecting a Style`](https://www.scalatest.org/user_guide/selecting_a_style) and applied the offending `Option.empty[String].contains(1234)` in each of the test blocks ([gist](https://gist.github.com/blast-hardcheese/e6790efa23c3b2ffbd14e5650f6a4875)). What I found was that not all tests exhibited this bug:

| Bug | No Bug | Partial Bug |
| ---- | ---- | ---- |
| `AnyFlatSpec` | `AnyFunSuite` |  |
| `AnyWordSpec` | `AnyFunSpec` |  |
| `AnyFreeSpec` |  | `AnyPropSpec` |

Resolution
---

I think a reasonable course of action here would be to sprinkle some unbounded type parameters on the functions that currently accept `Unit`. Unfortunately, my tests show that it's not possible to have both `def should[A](fun: => A): Unit` and `def should(fun: => Unit): Unit` visible at the same time, so care will need to be taken to ensure bincompat.

**Comment by blast-hardcheese** [comment]
This change would also resolve #1717, though that's a different situation.

**Comment by cheeseng** [comment]
@blast-hardcheese Interesting, thanks for the detailed analysis, I'll look at the style traits to try to understand why some have problem while some do not.

**Comment by cheeseng** [comment]
@blast-hardcheese I did a quick try and your workaround did work, but ideally, imho this is best fixed in the scala compiler, hopefully, it will be fixed, :)

**Comment by bvenners** [comment]
Rather than an unbounded and unused type parameter, I wonder if changing Unit to Any would solve the problem. We've done that in other places, so maybe we missed a few spots.

**Comment by blast-hardcheese** [comment]
@bvenners 
> I wonder if changing Unit to Any would solve the problem

I don't believe so -- in my reproduction repo, both cases are actually represented:

- `should(func: Unit): Unit`
- `println(x: Any): Unit`

That being said, nothing wrong with being explicit, so I updated the sample repo to include what I believe to be representative of your suggestion: https://github.com/blast-hardcheese/scala-bug-xlint-infer-any/commit/cd131ea242d0519b39966847233d1c4a9034297b
